### ISSUE TITLE
mgr pod in CrashLoop in 0.8.x

### ISSUE THREAD

**Issue by bjornryden** [rook/rook#2174]
**Is this a bug report or feature request?**
Bug Report

**Deviation from expected behavior:**
After rescheduling the mgr pod, it goes into a CrashLoop with the following:
```
2018-09-30 19:26:48.956022 I | ceph-mgr: 2018-09-30 19:26:48.955809 7f0adebf4700  1 mgr send_beacon active
2018-09-30 19:26:50.970860 I | ceph-mgr: 2018-09-30 19:26:50.970649 7f0adebf4700  1 mgr send_beacon active
2018-09-30 19:26:52.985827 I | ceph-mgr: 2018-09-30 19:26:52.985611 7f0adebf4700  1 mgr send_beacon active
2018-09-30 19:26:54.004538 I | ceph-mgr: [30/Sep/2018:19:26:47] ENGINE Bus STARTING
2018-09-30 19:26:54.004566 I | ceph-mgr: CherryPy Checker:
2018-09-30 19:26:54.004575 I | ceph-mgr: The Application mounted at '' has an empty config.
2018-09-30 19:26:54.004581 I | ceph-mgr: 
2018-09-30 19:26:54.004588 I | ceph-mgr: [30/Sep/2018:19:26:47] ENGINE Started monitor thread '_TimeoutMonitor'.
2018-09-30 19:26:54.004594 I | ceph-mgr: [30/Sep/2018:19:26:47] ENGINE Bus STARTING
2018-09-30 19:26:54.004600 I | ceph-mgr: [30/Sep/2018:19:26:47] ENGINE Started monitor thread '_TimeoutMonitor'.
2018-09-30 19:26:54.004606 I | ceph-mgr: [30/Sep/2018:19:26:47] ENGINE Serving on :::7000
2018-09-30 19:26:54.004611 I | ceph-mgr: [30/Sep/2018:19:26:47] ENGINE Bus STARTED
2018-09-30 19:26:54.004624 I | ceph-mgr: [30/Sep/2018:19:26:47] ENGINE Serving on :::9283
2018-09-30 19:26:54.004630 I | ceph-mgr: [30/Sep/2018:19:26:47] ENGINE Bus STARTED
2018-09-30 19:26:54.004636 I | ceph-mgr: terminate called after throwing an instance of 'std::out_of_range'
2018-09-30 19:26:54.004644 I | ceph-mgr:   what():  map::at
failed to run mgr. failed to start mgr: Failed to complete 'ceph-mgr': signal: aborted (core dumped).
```

**Expected behavior:**
No crash loop ;)

**How to reproduce it (minimal and precise):**
Personally I've experienced it in several test clusters, but haven't had time to dig into it until tonight.
Another user mentioned this a week ago on slack and @galexrt pointed to this bug in Ceph that seems to be related:
https://tracker.ceph.com/issues/24982

In that issue, people mention multiple RGWs and I'm running RGWs as a daemonset for these clusters.
So I tried to scale the amount of RGWs down to 1, using NodeAffinity, and nor the mgr was able to start up.
After it's started, I can scale the RGWs back up to full count (5 on testing cluster) in one go and the mgr stays up.
Without knowing this in depth, it seems to me the RGWs are building up a history of metrics to deliver to the mgr when they can't reach it. When the mgr starts again, these historic metrics overwhelms it and it gets startled, not to confuse with started.

**Environment**:
* OS (e.g. from /etc/os-release):
CoreOS
* Kernel (e.g. `uname -a`):
Something new
* Cloud provider or hardware configuration:
Bare metal
* Rook version (use `rook version` inside of a Rook Pod):
v0.8.1 - 99% sure I saw it on 0.8.2 as well, but had to vacate that due to a different bug.
* Kubernetes version (use `kubectl version`):
1.11.3
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift):
Kubespray
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/Rook/master/toolbox.html)):
Seems happy, but no dashboard and metrics obviously.

**Comment by soakes** [comment]
I can confirm that I also expernce this same issue but didnt manage to track it down to RGW's causing the problem until Bjorn kindly contacted me on Slack last night.  

The only solution I had was to switch to 0.7.1 for now which has been stable as a rock for many months on other clusters.

I will give this another whirl just with cephfs+rbd which hopefully should cure this (for now).  Thankfully I can get away without rgw for now at least on this new cluster.

**Comment by dimm0** [comment]
Same here right after deploying an object storage additional to block storage

```
2018-10-02 18:43:06.728867 I | ceph-mgr: 2018-10-02 18:43:06.728796 7f50a8b70700  1 mgr send_beacon active
2018-10-02 18:43:08.297400 I | ceph-mgr: ::ffff:10.244.16.163 - - [02/Oct/2018:18:43:08] "GET /metrics HTTP/1.1" 200 794936 "" "Prometheus/2.3.2"
2018-10-02 18:43:08.741118 I | ceph-mgr: 2018-10-02 18:43:08.741051 7f50a8b70700  1 mgr send_beacon active
2018-10-02 18:43:10.370099 I | ceph-mgr: [02/Oct/2018:18:43:04] ENGINE Bus STARTING
2018-10-02 18:43:10.370118 I | ceph-mgr: CherryPy Checker:
2018-10-02 18:43:10.370122 I | ceph-mgr: The Application mounted at '' has an empty config.
2018-10-02 18:43:10.370125 I | ceph-mgr:
2018-10-02 18:43:10.370128 I | ceph-mgr: [02/Oct/2018:18:43:04] ENGINE Started monitor thread '_TimeoutMonitor'.
2018-10-02 18:43:10.370130 I | ceph-mgr: terminate called after throwing an instance of 'std::out_of_range'
2018-10-02 18:43:10.370132 I | ceph-mgr:   what():  map::at
```

**Comment by galexrt** [comment]
@dmick @liewegas @rootfs I have the feeling this could be because of the "applications" we assign to the pools during the creation process by the Rook operator. Could that be an issue?

**Comment by dmick** [comment]
I don't think Cherrypy's terminology is related to Ceph's.  There doesn't seem to be enough information in the log to tell which manager plugin caused that issue; prometheus and dashboard both use it.

**Comment by dmick** [comment]
I'd probably try increasing the manager debug level.  "debug mgr = 10" or so in a ceph.conf, or a conf override, or whatever the easiest thing is

**Comment by dimm0** [comment]
Are you saying that it works fine in your test environment and you can't reproduce?

**Comment by bjornryden** [comment]
Worth noting: If you follow the thread @ ceph (https://tracker.ceph.com/issues/24982), people specifically note that it's a prometheus/metrics problem. It seems disabling metrics being pushed from the rgw's to the mgr alleviates the situation.
There's no merged fix it seems?

**Comment by dimm0** [comment]
Is there a way to disable metrics for S3, and only have blockdev working? (like it worked before enabling s3)

**Comment by elthariel** [comment]
Having the same issue on a RKE cluster running K8s 1.11.1

It's been working happily for 20 days, then a node reboot had the mgr re-scheduled on another box, then this crash loop started.

**Comment by travisn** [comment]
@elthariel if you scale the RGW services back to one, does the mgr start after that? The issue described in https://tracker.ceph.com/issues/24982 sounds like scaling the RGW services down temporarily should help workaround this.

**Comment by dimm0** [comment]
Hmm.. I modified the s3 objectstore operator to scale RGW to one, and it deleted the deployment and never recreated. Is there a way to create one again?

**Comment by dimm0** [comment]
Got it working with 1 RGW

**Comment by travisn** [comment]
@dimm0 What got it working again? You had to wait for the cluster to get healthy again before the operator would start rgw?

**Comment by dimm0** [comment]
Healthy cluster is not what it wanted. I had to manually clear some configmaps for non-existent nodes and kill prepare jobs which were not going to run for various reasons, so that the operator init would succeed (it's taking more than an hour on our cluster), only after that it created the RGW

I was going to suggest, could you separate different processes, which are independent of each other, into different goroutines and have them go in parallel? As I see it there's no real reason to wait for nodes provisioning step before deploying the RGW deployment..

**Comment by elthariel** [comment]
@travisn It's also working on my cluster when I scale the rgw to 1

**Comment by travisn** [comment]
@dimm0 agreed, there is no need for rgw to wait for osd orchestration. We've just kept the operator mostly single-threaded to keep it simple. I'd like to understand better why it takes so long for the orchestration to complete. It's taking a hour to process all the OSDs? Can you see anything in the logs that is slowing down the OSDs, or there are just that many OSDs?

**Comment by dimm0** [comment]
We have 180 OSDs currently, and going to have even more.

**Comment by galexrt** [comment]
@dimm0 As an idea, could you use the https://github.com/digitalocean/ceph_exporter as a "workaround" for metrics till the RGW + MGR prom module issue is fixed? (and disable the Prometheus module using a CronJo (operator will enable the module again on every start)).

**Comment by dimm0** [comment]
@galexrt hmm... would I need to rebuild the grafana dashboards?

**Comment by bjornryden** [comment]
> @galexrt hmm... would I need to rebuild the grafana dashboards?

From my vague memory, some metrics changes paths and some are added/removed. I think overall the metrics from the 3rd party exporter had more information than the build in one, but it's a year back since I looked at it.
I'm happy to rebuild the Grafana dashboard if it's just providing the json. If it's "scripted" dashboard, it's a bit beyond me at the moment.

**Comment by galexrt** [comment]
@dimm0 @bjornryden I think my dashboards were once compatible with the ceph exporter, so you can take a lucky guess and use an older revision from Grafana.com:
* "Ceph - Cluster" https://grafana.com/dashboards/2842/revisions
* "Ceph - OSD" https://grafana.com/dashboards/5336/revisions
* "Ceph - Pools" https://grafana.com/dashboards/5342/revisions

**Comment by dimm0** [comment]
At this point we're good with a single mgr

**Comment by JoshuaAndrew** [comment]
Same Problem !
```
2018-11-23 10:26:54.971459 I | rookcmd: starting Rook v0.8.3 with arguments '/usr/local/bin/rook ceph mgr --config-dir=/var/lib/rook'
2018-11-23 10:26:54.972324 I | rookcmd: flag values: --admin-secret=*****, --ceph-config-override=/etc/rook/config/override.conf, --cluster-name=rook-ceph, --config-dir=/var/lib/rook, --fsid=, --help=false, --log-level=INFO, --mgr-keyring=AQA7rvdbqPJyHRAADXReLAcziNtUeCMMMcuBeg==, --mgr-name=a, --mon-endpoints=rook-ceph-mon0=10.103.141.47:6790,rook-ceph-mon1=10.102.177.14:6790,rook-ceph-mon2=10.100.205.231:6790, --mon-secret=*****, --private-ip=172.31.72.14, --private-ipv4=, --public-ip=172.31.72.14, --public-ipv4=
2018-11-23 10:26:54.972334 I | cephmon: parsing mon endpoints: rook-ceph-mon0=10.103.141.47:6790,rook-ceph-mon1=10.102.177.14:6790,rook-ceph-mon2=10.100.205.231:6790
2018-11-23 10:26:54.972349 I | cephmgr: Starting MGR a with keyring AQA7rvdbqPJyHRAADXReLAcziNtUeCMMMcuBeg==
2018-11-23 10:26:54.972358 I | cephmgr: Conf files: dir=/var/lib/rook/mgr-a keyring=/var/lib/rook/mgr-a/keyring
2018-11-23 10:26:54.972603 I | cephmon: writing config file /var/lib/rook/mgr-a/rook-ceph.config
2018-11-23 10:26:54.972873 I | cephmon: copying config to /etc/ceph/ceph.conf
2018-11-23 10:26:54.973472 I | cephmgr: starting ceph-mgr
2018-11-23 10:26:54.973548 I | cephmgr: Config file /var/lib/rook/mgr-a/rook-ceph.config:
[global]
run dir                   = /var/lib/rook/mgr-a
mon initial members       = rook-ceph-mon0 rook-ceph-mon1 rook-ceph-mon2
mon host                  = 10.103.141.47:6790,10.102.177.14:6790,10.100.205.231:6790
log file                  = /dev/stdout
mon cluster log file      = /dev/stdout
public addr               = 172.31.72.14
cluster addr              = 172.31.72.14
mon keyvaluedb            = rocksdb
mon_allow_pool_delete     = true
mon_max_pg_per_osd        = 1000
debug default             = 0
debug rados               = 0
debug mon                 = 0
debug osd                 = 0
debug bluestore           = 0
debug filestore           = 0
debug journal             = 0
debug leveldb             = 0
filestore_omap_backend    = rocksdb
osd pg bits               = 11
osd pgp bits              = 11
osd pool default size     = 1
osd pool default min size = 1
osd pool default pg num   = 100
osd pool default pgp num  = 100
rbd_default_features      = 3
fatal signal handlers     = false

[mgr.a]
keyring  = /var/lib/rook/mgr-a/keyring
mgr data = /var/lib/rook/mgr-a

2018-11-23 10:26:54.973581 I | cephmgr: Config file /var/lib/rook/mgr-a/keyring:

[mgr.a]
	key = AQA7rvdbqPJyHRAADXReLAcziNtUeCMMMcuBeg==
	caps mon = "allow *"
2018-11-23 10:26:54.973598 I | exec: Running command: ceph-mgr --foreground --cluster=rook-ceph --conf=/var/lib/rook/mgr-a/rook-ceph.config --keyring=/var/lib/rook/mgr-a/keyring -i a
2018-11-23 10:26:55.025856 I | ceph-mgr: 2018-11-23 10:26:55.025592 7f04a9de6680  0 ceph version 12.2.7 (3ec878d1e53e1aeb47a9f619c49d9e7c0aa384d5) luminous (stable), process ceph-mgr, pid 14
2018-11-23 10:26:55.025903 I | ceph-mgr: 2018-11-23 10:26:55.025867 7f04a9de6680  0 pidfile_write: ignore empty --pid-file
2018-11-23 10:26:55.067308 I | ceph-mgr: 2018-11-23 10:26:55.067197 7f04a9de6680  1 mgr send_beacon standby
2018-11-23 10:26:55.098491 I | ceph-mgr: 2018-11-23 10:26:55.098355 7f04a1149700  1 mgr init Loading python module 'balancer'
2018-11-23 10:26:55.172333 I | ceph-mgr: 2018-11-23 10:26:55.172253 7f04a1149700  1 mgr init Loading python module 'dashboard'
2018-11-23 10:26:55.370527 I | ceph-mgr: 2018-11-23 10:26:55.370453 7f04a1149700  1 mgr init Loading python module 'prometheus'
2018-11-23 10:26:55.451099 I | ceph-mgr: 2018-11-23 10:26:55.451013 7f04a1149700  1 mgr init Loading python module 'restful'
2018-11-23 10:26:55.671761 I | ceph-mgr: 2018-11-23 10:26:55.671678 7f04a1149700  1 mgr init Loading python module 'status'
2018-11-23 10:26:55.742692 I | ceph-mgr: 2018-11-23 10:26:55.742163 7f04a1149700  1 mgr handle_mgr_map Activating!
2018-11-23 10:26:55.742717 I | ceph-mgr: 2018-11-23 10:26:55.742605 7f04a1149700  1 mgr handle_mgr_map I am now activating
2018-11-23 10:26:55.751828 I | ceph-mgr: 2018-11-23 10:26:55.751759 7f048ae10700  1 mgr load Constructed class from module: balancer
2018-11-23 10:26:55.752539 I | ceph-mgr: 2018-11-23 10:26:55.752463 7f048ae10700  1 mgr load Constructed class from module: dashboard
2018-11-23 10:26:55.753194 I | ceph-mgr: 2018-11-23 10:26:55.753128 7f048ae10700  1 mgr load Constructed class from module: prometheus
2018-11-23 10:26:55.753466 I | ceph-mgr: 2018-11-23 10:26:55.753412 7f048ae10700  1 mgr load Constructed class from module: restful
2018-11-23 10:26:55.753881 I | ceph-mgr: 2018-11-23 10:26:55.753832 7f048ae10700  1 mgr load Constructed class from module: status
2018-11-23 10:26:55.754263 I | ceph-mgr: 2018-11-23 10:26:55.753926 7f048ae10700  1 mgr send_beacon active
2018-11-23 10:26:55.755612 I | ceph-mgr: 2018-11-23 10:26:55.755545 7f0487e0a700  1 mgr[restful] server not running: no certificate configured
2018-11-23 10:26:55.756930 I | ceph-mgr: 2018-11-23 10:26:55.756867 7f048ae10700  1 mgr[restful] Unknown request ''
2018-11-23 10:26:55.757664 I | ceph-mgr: 2018-11-23 10:26:55.757532 7f048ae10700  1 mgr[restful] Unknown request ''
2018-11-23 10:26:56.777828 I | ceph-mgr: 2018-11-23 10:26:56.776868 7f04a394e700  0 -- 172.31.72.14:6800/14 >> 172.31.241.130:6800/15200 conn(0x55b6b57d1800 :6800 s=STATE_ACCEPTING_WAIT_CONNECT_MSG_AUTH pgs=0 cs=0 l=1).handle_connect_msg: challenging authorizer
2018-11-23 10:26:56.777854 I | ceph-mgr: 2018-11-23 10:26:56.777006 7f04a394e700  0 -- 172.31.72.14:6800/14 >> 172.31.241.128:0/36 conn(0x55b6b57d0000 :6800 s=STATE_ACCEPTING_WAIT_CONNECT_MSG_AUTH pgs=0 cs=0 l=1).handle_connect_msg: challenging authorizer
2018-11-23 10:26:56.777859 I | ceph-mgr: 2018-11-23 10:26:56.777243 7f04a414f700  0 -- 172.31.72.14:6800/14 >> 172.31.241.193:6800/12567 conn(0x55b6b57d3000 :6800 s=STATE_ACCEPTING_WAIT_CONNECT_MSG_AUTH pgs=0 cs=0 l=1).handle_connect_msg: challenging authorizer
2018-11-23 10:26:56.778263 I | ceph-mgr: 2018-11-23 10:26:56.778185 7f04a394e700  0 -- 172.31.72.14:6800/14 >> 172.31.222.180:6800/303711 conn(0x55b6b57e3000 :6800 s=STATE_ACCEPTING_WAIT_CONNECT_MSG_AUTH pgs=0 cs=0 l=1).handle_connect_msg: challenging authorizer
2018-11-23 10:26:56.778609 I | ceph-mgr: 2018-11-23 10:26:56.778537 7f04a414f700  0 -- 172.31.72.14:6800/14 >> 172.31.72.52:6800/514212 conn(0x55b6b57e1800 :6800 s=STATE_ACCEPTING_WAIT_CONNECT_MSG_AUTH pgs=0 cs=0 l=1).handle_connect_msg: challenging authorizer
2018-11-23 10:26:56.782109 I | ceph-mgr: 2018-11-23 10:26:56.782043 7f04a414f700  0 -- 172.31.72.14:6800/14 >> 172.31.141.208:6800/1448 conn(0x55b6b5873000 :6800 s=STATE_ACCEPTING_WAIT_CONNECT_MSG_AUTH pgs=0 cs=0 l=1).handle_connect_msg: challenging authorizer
2018-11-23 10:26:56.782223 I | ceph-mgr: 2018-11-23 10:26:56.782167 7f04a414f700  0 -- 172.31.72.14:6800/14 >> 172.31.104.86:6800/125294 conn(0x55b6b57fa800 :6800 s=STATE_ACCEPTING_WAIT_CONNECT_MSG_AUTH pgs=0 cs=0 l=1).handle_connect_msg: challenging authorizer
2018-11-23 10:26:56.783622 I | ceph-mgr: 2018-11-23 10:26:56.783538 7f04a314d700  0 -- 172.31.72.14:6800/14 >> 172.31.176.16:0/3131600053 conn(0x55b6b57e0000 :6800 s=STATE_ACCEPTING_WAIT_CONNECT_MSG_AUTH pgs=0 cs=0 l=1).handle_connect_msg: challenging authorizer
2018-11-23 10:26:56.784555 I | ceph-mgr: 2018-11-23 10:26:56.784456 7f04a414f700  0 -- 172.31.72.14:6800/14 >> 172.31.139.33:6800/9424 conn(0x55b6b582e000 :6800 s=STATE_ACCEPTING_WAIT_CONNECT_MSG_AUTH pgs=0 cs=0 l=1).handle_connect_msg: challenging authorizer
2018-11-23 10:26:56.784753 I | ceph-mgr: 2018-11-23 10:26:56.784652 7f04a314d700  0 -- 172.31.72.14:6800/14 >> 172.31.176.16:0/684700236 conn(0x55b6b5886000 :6800 s=STATE_ACCEPTING_WAIT_CONNECT_MSG_AUTH pgs=0 cs=0 l=1).handle_connect_msg: challenging authorizer
2018-11-23 10:26:56.794867 I | ceph-mgr: 2018-11-23 10:26:56.783815 7f04a394e700  0 -- 172.31.72.14:6800/14 >> 172.31.176.1:6800/3010 conn(0x55b6b5870000 :6800 s=STATE_ACCEPTING_WAIT_CONNECT_MSG_AUTH pgs=0 cs=0 l=1).handle_connect_msg: challenging authorizer
2018-11-23 10:26:56.794891 I | ceph-mgr: 2018-11-23 10:26:56.784815 7f04a394e700  0 -- 172.31.72.14:6800/14 >> 172.31.139.31:0/36 conn(0x55b6b5874800 :6800 s=STATE_ACCEPTING_WAIT_CONNECT_MSG_AUTH pgs=0 cs=0 l=1).handle_connect_msg: challenging authorizer
2018-11-23 10:26:56.794897 I | ceph-mgr: 2018-11-23 10:26:56.789547 7f04a314d700  0 -- 172.31.72.14:6800/14 >> 172.31.222.185:0/926881167 conn(0x55b6b5871800 :6800 s=STATE_ACCEPTING_WAIT_CONNECT_MSG_AUTH pgs=0 cs=0 l=1).handle_connect_msg: challenging authorizer
2018-11-23 10:26:56.794900 I | ceph-mgr: 2018-11-23 10:26:56.789702 7f04a394e700  0 -- 172.31.72.14:6800/14 >> 172.31.55.158:6800/6127 conn(0x55b6b582f800 :6800 s=STATE_ACCEPTING_WAIT_CONNECT_MSG_AUTH pgs=0 cs=0 l=1).handle_connect_msg: challenging authorizer
2018-11-23 10:26:56.794903 I | ceph-mgr: 2018-11-23 10:26:56.789919 7f04a394e700  0 -- 172.31.72.14:6800/14 >> 172.31.222.185:0/384063172 conn(0x55b6b5887800 :6800 s=STATE_ACCEPTING_WAIT_CONNECT_MSG_AUTH pgs=0 cs=0 l=1).handle_connect_msg: challenging authorizer
2018-11-23 10:26:56.801607 I | ceph-mgr: 2018-11-23 10:26:56.801529 7f04a314d700  0 -- 172.31.72.14:6800/14 >> 172.31.109.81:6800/13942 conn(0x55b6b58ee000 :6800 s=STATE_ACCEPTING_WAIT_CONNECT_MSG_AUTH pgs=0 cs=0 l=1).handle_connect_msg: challenging authorizer
2018-11-23 10:26:56.802386 I | ceph-mgr: 2018-11-23 10:26:56.802278 7f04a394e700  0 -- 172.31.72.14:6800/14 >> 172.31.140.115:0/1390181397 conn(0x55b6b58ef800 :6800 s=STATE_ACCEPTING_WAIT_CONNECT_MSG_AUTH pgs=0 cs=0 l=1).handle_connect_msg: challenging authorizer
2018-11-23 10:26:56.807156 I | ceph-mgr: 2018-11-23 10:26:56.802932 7f04a414f700  0 -- 172.31.72.14:6800/14 >> 172.31.140.115:0/3688425393 conn(0x55b6b5832800 :6800 s=STATE_ACCEPTING_WAIT_CONNECT_MSG_AUTH pgs=0 cs=0 l=1).handle_connect_msg: challenging authorizer
2018-11-23 10:26:56.807169 I | ceph-mgr: 2018-11-23 10:26:56.805744 7f04a394e700  0 -- 172.31.72.14:6800/14 >> 172.31.140.113:6800/28361 conn(0x55b6b58f2800 :6800 s=STATE_ACCEPTING_WAIT_CONNECT_MSG_AUTH pgs=0 cs=0 l=1).handle_connect_msg: challenging authorizer
2018-11-23 10:26:56.807174 I | ceph-mgr: 2018-11-23 10:26:56.805959 7f04a414f700  0 -- 172.31.72.14:6800/14 >> 172.31.109.79:0/36 conn(0x55b6b58f1000 :6800 s=STATE_ACCEPTING_WAIT_CONNECT_MSG_AUTH pgs=0 cs=0 l=1).handle_connect_msg: challenging authorizer
2018-11-23 10:26:56.809435 I | ceph-mgr: 2018-11-23 10:26:56.809291 7f04a314d700  0 -- 172.31.72.14:6800/14 >> 172.31.76.154:6800/9397 conn(0x55b6b593e000 :6800 s=STATE_ACCEPTING_WAIT_CONNECT_MSG_AUTH pgs=0 cs=0 l=1).handle_connect_msg: challenging authorizer
2018-11-23 10:26:56.818539 I | ceph-mgr: 2018-11-23 10:26:56.818456 7f04a394e700  0 -- 172.31.72.14:6800/14 >> 172.31.86.156:6800/30358 conn(0x55b6b58e8800 :6800 s=STATE_ACCEPTING_WAIT_CONNECT_MSG_AUTH pgs=0 cs=0 l=1).handle_connect_msg: challenging authorizer
2018-11-23 10:26:57.070352 I | ceph-mgr: 2018-11-23 10:26:57.070200 7f049e143700  1 mgr send_beacon active
2018-11-23 10:26:59.071517 I | ceph-mgr: 2018-11-23 10:26:59.071395 7f049e143700  1 mgr send_beacon active
2018-11-23 10:27:01.072444 I | ceph-mgr: 2018-11-23 10:27:01.072278 7f049e143700  1 mgr send_beacon active
2018-11-23 10:27:01.918829 I | ceph-mgr: [23/Nov/2018:10:26:55] ENGINE Bus STARTING
2018-11-23 10:27:01.918855 I | ceph-mgr: [23/Nov/2018:10:26:55] ENGINE Started monitor thread '_TimeoutMonitor'.
2018-11-23 10:27:01.918859 I | ceph-mgr: [23/Nov/2018:10:26:55] ENGINE Bus STARTING
2018-11-23 10:27:01.918863 I | ceph-mgr: CherryPy Checker:
2018-11-23 10:27:01.918867 I | ceph-mgr: The Application mounted at '' has an empty config.
2018-11-23 10:27:01.918869 I | ceph-mgr: 
2018-11-23 10:27:01.918877 I | ceph-mgr: [23/Nov/2018:10:26:55] ENGINE Started monitor thread '_TimeoutMonitor'.
2018-11-23 10:27:01.918881 I | ceph-mgr: [23/Nov/2018:10:26:55] ENGINE Serving on :::7000
2018-11-23 10:27:01.918884 I | ceph-mgr: [23/Nov/2018:10:26:55] ENGINE Bus STARTED
2018-11-23 10:27:01.918887 I | ceph-mgr: [23/Nov/2018:10:26:55] ENGINE Serving on :::9283
2018-11-23 10:27:01.918890 I | ceph-mgr: [23/Nov/2018:10:26:55] ENGINE Bus STARTED
2018-11-23 10:27:01.918893 I | ceph-mgr: terminate called after throwing an instance of 'std::out_of_range'
2018-11-23 10:27:01.918896 I | ceph-mgr:   what():  map::at
failed to run mgr. failed to start mgr: Failed to complete 'ceph-mgr': signal: aborted (core dumped). 
```

**Comment by dimm0** [comment]
What's needed to fix this? Would new ceph version have the fix?
@galexrt @travisn

**Comment by galexrt** [comment]
@dimm0 Possibly yes, because you can choose the Ceph version in master and v0.9 that is running, it isn't a problem anymore to change the Ceph version.
So in the "worst" case we simply need to advice people to use a more up-to-date Ceph version (for upgrades from v0.8 to v0.9 Ceph Luminous (latest release I think which could potentially already fix the issues posted about here) is used AFAIK).

**Comment by dimm0** [comment]
Would you have the update then in place for 0.9?

**Comment by galexrt** [comment]
@dimm0 What update?

**Comment by dimm0** [comment]
yup

**Comment by galexrt** [comment]
@dimm0 Yes it is included, as you can see in the https://github.com/rook/rook/blob/master/PendingReleaseNotes.md#notable-features the first point.

**Comment by dimm0** [comment]
Still an issue in rook 0.9, rgw & mgr v13.2.2-20181023

**Comment by christianhuening** [comment]
seeing the same here with rook v0.9.0-19.g8f94838 and CEPH v13.2.2-20181023

**Comment by Amos-85** [comment]
same issue in rook 0.9.1 / v0.9.0-60.gde8e141 & ceph v13.2.4-20190109
multiple rgw instances cause the mgr to crash.

**Comment by travisn** [comment]
@dimm0 According to the [last comment in the ceph issue](https://tracker.ceph.com/issues/24982#note-15), the following setting in [ceph.conf](https://rook.io/docs/rook/v0.9/advanced-configuration.html#custom-cephconf-settings) may help. Can you see if this helps your cluster? It looks like you'll also need luminous 12.2.9 or newer, but I can't tell what version of mimic might contain a fix for sure.
```
mgr_stats_threshold = 11
```

**Comment by dimm0** [comment]
Hm... The mgr is not crashing, but some stats, like IOPS, throughput and operations, are not reported. Is that how it's supposed to work?

**Comment by dimm0** [comment]
Thanks! I'll try updating to 13.2.4. I'm currently on 13.2.2

**Comment by christianhuening** [comment]
@Amos-85 i am on the same version now and for me the issue still persists. Multiple RGW's are causing the mgr to die:
```
  ~ kubectl -n rook-ceph logs rook-ceph-mgr-a-57cd774d86-78bsn

2019-01-21 15:37:06.195 7f79374fc380  0 ceph version 13.2.4 (b10be4d44915a4d78a8e06aa31919e74927b142e) mimic (stable), process ceph-mgr, pid 1
2019-01-21 15:37:06.195 7f79374fc380  0 pidfile_write: ignore empty --pid-file
2019-01-21 15:37:06.231 7f79374fc380  1 mgr[py] Loading python module 'balancer'
2019-01-21 15:37:06.255 7f79374fc380  1 mgr[py] Loading python module 'dashboard'
2019-01-21 15:37:06.339 7f79374fc380  1 mgr[py] Loading python module 'hello'
2019-01-21 15:37:06.353 7f79374fc380  1 mgr[py] Loading python module 'influx'
2019-01-21 15:37:06.365 7f79374fc380  1 mgr[py] Loading python module 'iostat'
2019-01-21 15:37:06.379 7f79374fc380  1 mgr[py] Loading python module 'localpool'
2019-01-21 15:37:06.391 7f79374fc380  1 mgr[py] Loading python module 'prometheus'
2019-01-21 15:37:06.446 7f79374fc380  1 mgr[py] Loading python module 'restful'
2019-01-21 15:37:06.603 7f79374fc380  1 mgr[py] Loading python module 'selftest'
2019-01-21 15:37:06.617 7f79374fc380  1 mgr[py] Loading python module 'smart'
2019-01-21 15:37:06.630 7f79374fc380  1 mgr[py] Loading python module 'status'
2019-01-21 15:37:06.655 7f79374fc380  1 mgr[py] Loading python module 'telegraf'
2019-01-21 15:37:06.678 7f79374fc380  1 mgr[py] Loading python module 'telemetry'
2019-01-21 15:37:06.765 7f79374fc380  1 mgr[py] Loading python module 'zabbix'
2019-01-21 15:37:06.783 7f79374fc380  1 mgr send_beacon standby
2019-01-21 15:37:06.845 7f7923355700  1 mgr handle_mgr_map Activating!
2019-01-21 15:37:06.845 7f7923355700  1 mgr handle_mgr_map I am now activating
2019-01-21 15:37:06.854 7f7918205700  1 mgr load Constructed class from module: balancer
2019-01-21 15:37:06.855 7f7918205700  1 mgr load Constructed class from module: dashboard
2019-01-21 15:37:06.856 7f7918205700  1 mgr load Constructed class from module: prometheus
2019-01-21 15:37:06.857 7f7918205700  1 mgr load Constructed class from module: restful
2019-01-21 15:37:06.857 7f7918205700  1 mgr load Constructed class from module: status
2019-01-21 15:37:06.857 7f7918205700  1 mgr send_beacon active
[21/Jan/2019:15:37:06] ENGINE Bus STARTING
CherryPy Checker:
The Application mounted at '' has an empty config.

2019-01-21 15:37:06.860 7f79151ff700  1 mgr[restful] server not running: no certificate configured
[21/Jan/2019:15:37:06] ENGINE Started monitor thread '_TimeoutMonitor'.
[21/Jan/2019:15:37:06] ENGINE Serving on 10.6.64.99:9283
[21/Jan/2019:15:37:06] ENGINE Bus STARTED
[21/Jan/2019:15:37:06] ENGINE Bus STARTING
[21/Jan/2019:15:37:06] ENGINE Started monitor thread '_TimeoutMonitor'.
[21/Jan/2019:15:37:07] ENGINE Serving on 10.6.64.99:8443
[21/Jan/2019:15:37:07] ENGINE Bus STARTED
10.6.65.53 - - [21/Jan/2019:15:37:08] "GET /metrics HTTP/1.1" 200 95972 "" "Prometheus/2.4.3"
2019-01-21 15:37:08.784 7f791fb4e700  1 mgr send_beacon active
10.6.64.34 - - [21/Jan/2019:15:37:08] "GET /metrics HTTP/1.1" 200 95972 "" "Prometheus/2.4.3"
10.6.66.38 - - [21/Jan/2019:15:37:10] "GET /metrics HTTP/1.1" 200 95972 "" "Prometheus/2.4.3"
2019-01-21 15:37:10.786 7f791fb4e700  1 mgr send_beacon active
2019-01-21 15:37:12.787 7f791fb4e700  1 mgr send_beacon active
terminate called after throwing an instance of 'std::out_of_range'
  what():  map::at
```

**Comment by Amos-85** [comment]
@christianhuening  ,
what is the rook operator docker image & chart version you are using?

**Comment by christianhuening** [comment]
@Amos-85 I am using v0.9.0-79.g1a1ffdd 
Chart is latest stable

**Comment by patrickstjohn** [comment]
@christianhuening something else you may want to try that we have with our setup is using the Beast RGW frontend instead of the standard Civetweb. We've gotten better performance with our gateway's with the newer frontend. 

http://docs.ceph.com/docs/master/radosgw/config-ref/
http://docs.ceph.com/docs/master/radosgw/frontends/

Otherwise, our primary difference from what I think you are running is we're on the latest v0.9.1 instead of v0.9.0 and Ceph 13.2.4.

**Comment by christianhuening** [comment]
@patrickstjohn thanks! I've been on 0.9.1, but we required TLS on RGW to be working, so i switched over to the latest build, which contains a fix for the TLS mount. 

Can you quantify the performance gain you get from Beast?

**Comment by dimm0** [comment]
Updated to v13.2.4-20190109, still the same problem

**Comment by patrickstjohn** [comment]
@christianhuening We were seeing roughly 2Gbit/sec on our r/w throughput with a tuned civetweb and with the beast frontend we are getting 3Gbit/sec throughput.

Additionally, we're deploying with 12 RGW's off the bat. We are not deploying with 1 and then scaling it to 12. I don't know if that differs from anyone else's deployment model.

**Comment by christianhuening** [comment]
Whether starting out with 1 and then scaling or directly starting at scale doesn't matter for the crash. I tested that.

**Comment by Amos-85** [comment]
@christianhuening ,
I found something mysterious,
I had the same issue now,
3 rgw instances and the mgr is failed,
with 1 rgw the mgr is function properly,

if I scale to 3 rgw & redploy them, then the mgr back to normal state.

**Comment by dimm0** [comment]
Fixed in mimic 13.2.5
This can be closed
